File: src/main/java/in/neuw/resource/demo/security/CustomBearerTokenServerAccessDeniedHandler.java
package in.neuw.resource.demo.security;

import com.google.gson.Gson;
import in.neuw.resource.demo.models.ErrorDto;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.buffer.DefaultDataBufferFactory;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.oauth2.server.resource.BearerTokenErrorCodes;
import org.springframework.security.oauth2.server.resource.authentication.AbstractOAuth2TokenAuthenticationToken;
import org.springframework.security.web.server.authorization.ServerAccessDeniedHandler;
import org.springframework.web.server.ServerWebExchange;
import reactor.core.publisher.Mono;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * @author Karanbir Singh on 04/21/2020
 */
public class CustomBearerTokenServerAccessDeniedHandler implements ServerAccessDeniedHandler {

    private String realmName;

    @Autowired
    public Gson gson;

    @Override
    public Mono<Void> handle(ServerWebExchange exchange, AccessDeniedException denied) {

        Map<String, String> parameters = new LinkedHashMap<>();

        if (this.realmName != null) {
            parameters.put("realm", this.realmName);
        }

        return exchange.getPrincipal()
                .filter(AbstractOAuth2TokenAuthenticationToken.class::isInstance)
                .map(token -> errorMessageParameters(parameters))
                .switchIfEmpty(Mono.just(parameters))
                .flatMap(params -> respond(exchange, params));
    }

    /**
     * Set the default realm name to use in the bearer token error response
     *
     * @param realmName
     */
    public final void setRealmName(String realmName) {
        this.realmName = realmName;
    }

    private static Map<String, String> errorMessageParameters(Map<String, String> parameters) {
        parameters.put("error", BearerTokenErrorCodes.INSUFFICIENT_SCOPE);
        parameters.put("error_description", "The request requires higher privileges than provided by the access token.");
        parameters.put("error_uri", "https://tools.ietf.org/html/rfc6750#section-3.1");

        return parameters;
    }

    private Mono<Void> respond(ServerWebExchange exchange, Map<String, String> parameters) {
        String wwwAuthenticate = computeWWWAuthenticateHeaderValue(parameters);
        exchange.getResponse().setStatusCode(HttpStatus.FORBIDDEN);
        exchange.getResponse().getHeaders().set(HttpHeaders.WWW_AUTHENTICATE, wwwAuthenticate);

        var errorDto = new ErrorDto(parameters.get("error_description"), HttpStatus.FORBIDDEN);
        errorDto.setCode(101);
        errorDto.setError(parameters.get("error"));

        return exchange.getResponse().writeWith(Mono.just(new DefaultDataBufferFactory().wrap(gson.toJson(errorDto).getBytes())));
    }

    private static String computeWWWAuthenticateHeaderValue(Map<String, String> parameters) {
        StringBuilder wwwAuthenticate = new StringBuilder();
        wwwAuthenticate.append("Bearer");
        if (!parameters.isEmpty()) {
            wwwAuthenticate.append(" ");
            int i = 0;
            for (Map.Entry<String, String> entry : parameters.entrySet()) {
                wwwAuthenticate.append(entry.getKey()).append("=\"").append(entry.getValue()).append("\"");
                if (i != parameters.size() - 1) {
                    wwwAuthenticate.append(", ");
                }
                i++;
            }
        }

        return wwwAuthenticate.toString();
    }
}

File: src/main/kotlin/in/neuw/resource/demo/security/CustomBearerTokenServerAccessDeniedHandler.kt
package `in`.neuw.resource.demo.security

import `in`.neuw.resource.demo.models.ErrorDto
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.core.io.buffer.DefaultDataBufferFactory
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.oauth2.server.resource.BearerTokenErrorCodes
import org.springframework.security.oauth2.server.resource.authentication.AbstractOAuth2TokenAuthenticationToken
import org.springframework.security.web.server.authorization.ServerAccessDeniedHandler
import org.springframework.stereotype.Component
import org.springframework.web.server.ServerWebExchange
import reactor.core.publisher.Mono
import java.security.Principal
import java.util.*

@Component
class CustomBearerTokenServerAccessDeniedHandler: ServerAccessDeniedHandler {

    @Autowired
    private lateinit var jacksonObjectMapper: ObjectMapper

    private var realmName: String? = null

    override fun handle(exchange: ServerWebExchange, denied: AccessDeniedException?): Mono<Void?>? {
        val parameters: MutableMap<String, String?> = LinkedHashMap()
        if (realmName != null) {
            parameters["realm"] = realmName
        }
        return exchange.getPrincipal<Principal>()
                .filter { obj: Principal? -> AbstractOAuth2TokenAuthenticationToken::class.java.isInstance(obj) }
                .map { token: Principal? -> errorMessageParameters(parameters) }
                .switchIfEmpty(Mono.just<Map<String, String?>>(parameters))
                .flatMap { params: Map<String, String?> -> respond(exchange, params) }
    }

    /**
     * Set the default realm name to use in the bearer token error response
     *
     * @param realmName
     */
    fun setRealmName(realmName: String?) {
        this.realmName = realmName
    }

    private fun errorMessageParameters(parameters: MutableMap<String, String?>): Map<String, String?> {
        parameters["error"] = BearerTokenErrorCodes.INSUFFICIENT_SCOPE
        parameters["error_description"] = "The request requires higher privileges than provided by the access token."
        parameters["error_uri"] = "https://tools.ietf.org/html/rfc6750#section-3.1"
        return parameters
    }

    private fun respond(exchange: ServerWebExchange, parameters: Map<String, String?>): Mono<Void?>? {
        return Mono.defer {
            val wwwAuthenticate = computeWWWAuthenticateHeaderValue(parameters)
            exchange.response.statusCode = HttpStatus.FORBIDDEN
            exchange.response.headers[HttpHeaders.WWW_AUTHENTICATE] = wwwAuthenticate

            var errorDto = ErrorDto(parameters.get("error_description"), HttpStatus.FORBIDDEN)
            errorDto.code = 101
            errorDto.error = parameters["error"]!!

            return@defer exchange.response.writeWith(Mono.just(
                    DefaultDataBufferFactory().wrap(jacksonObjectMapper.writeValueAsBytes(errorDto))))
        }
        //return exchange.response.setComplete()
    }

    private fun computeWWWAuthenticateHeaderValue(parameters: Map<String, String?>): String {
        val wwwAuthenticate = StringBuilder()
        wwwAuthenticate.append("Bearer")
        if (!parameters.isEmpty()) {
            wwwAuthenticate.append(" ")
            var i = 0
            for ((key, value) in parameters) {
                wwwAuthenticate.append(key).append("=\"").append(value).append("\"")
                if (i != parameters.size - 1) {
                    wwwAuthenticate.append(", ")
                }
                i++
            }
        }
        return wwwAuthenticate.toString()
    }

}